pkg/image/registryclient/client.go

```diff
@@ -374,6 +429,24 @@ func NewLimitedRetryRepository(repo distribution.Repository, retries int, limite
 	}
 }
 
+func (c *Context) reposForRef(ctx context.Context, locator imagereference.DockerImageReference, sources []imagereference.DockerImageReference) error {
+	cachedRepos := make(map[imagereference.DockerImageReference]distribution.Repository)
+	for _, altRef := range sources {
+		altRepo, err := c.Repository(ctx, altRef.DockerClientDefaults().RegistryURL(), altRef.RepositoryName(), true)
+		if err != nil {
+			return err
+		}
+		if altRepo != nil {
+			cachedRepos[altRef] = altRepo
+		}
+	}
+	if len(cachedRepos) == 0 {
+		return fmt.Errorf("did not find any repositories for %s", locator)
```

Comment:
I think it would be better here to return an error indicating that we have no more "alternates" to try. Something similar to what happens when we reach the end of a file (`io.EOF`).  That way one can easily distinguish between an expected error (there is no more alternates) than an unexpected one (like the one returned on line 435 above).

Reply:
the commit is outdated, as i updated/pushed thinking there were no more outstanding reviews - however, I updated the error message there to indicate "alternates" instead of an error fetching the repository.

Reply:
I see. The use case I am trying to tackle is the following:

1) try all alternate registry for a given manifest
2) if could not find the manifest in any of the alternate registries give the original registry a try

On such scenario I would iterate over `NextRepositoryWithAlternateReference()` until it returns an error and then move to the original registry but the only way I can distinguish between the end of the alternate registries list and any other error is by comparing the string within the error. That is what I would like to avoid. What I was proposing is an specific error that then I can compare, something like `if err == EndOfAlternates`.

If the proposed scenario is not a valid one then please let me know.

/cc @dmage

Reply:
Nevermind, this has been addressed by your last commits. I don't need to iterate anymore. Thanks.

Reply:
oh great! and yea, if it makes sense anywhere we'll add a custom error.